src/shared/storage/ws.ts

```diff
@@ -92,19 +103,63 @@ export class WebSocketBrowserStorageProvider implements StorageProvider {
           offset: req.payload.offset,
         });
       } else {
-        this.logger.error(
-          `[WebSocketBrowserStorageProvider] Unsupported message in publishData(): ${
-            (req as GsbRequest<void>).component
-          }`,
-        );
+        this.logger.error(`Unsupported message in publishData(): ${(req as GsbRequest<void>).component}`);
       }
     });
 
     return fileInfo.url;
   }
 
-  async publishFile(): Promise<string> {
-    throw new GolemInternalError("Not implemented");
+  async publishFile(src: string): Promise<string> {
+    if (isBrowser) {
+      throw new GolemInternalError("Cannot publish files in browser context, did you mean to use `publishData()`?");
+    }
+
+    this.logger.info("Preparing file upload", { sourcePath: src });
+
+    const fileInfo = await this.createFileInfo();
+    const ws = await this.createSocket(fileInfo, ["GetMetadata", "GetChunk"]);
+    const fileStats = await fsPromises.stat(src);
+    const fileSize = fileStats.size;
+
+    const fileHandle = await fsPromises.open(src, "r");
+    this.openHandles.add(fileHandle);
+
+    ws.addEventListener("message", async (event) => {
+      if (!(event.data instanceof ArrayBuffer)) {
+        this.logger.error("Received non-ArrayBuffer data from the socket", { data: event.data });
+        return;
+      }
+
+      const req = toObject(event.data) as GsbRequestPublishUnion;
+
+      if (req.component === "GetMetadata") {
+        this.respond(ws, req.id, { fileSize });
+      } else if (req.component === "GetChunk") {
+        const { offset, size } = req.payload;
+
+        const chunkSize = Math.min(size, fileSize - offset);
+        const chunk = Buffer.alloc(chunkSize);
+
+        try {
+          await fileHandle.read(chunk, 0, chunkSize, offset);
+          this.respond(ws, req.id, {
+            content: chunk,
+            offset,
+          });
+        } finally {
+          // After the last chunk, close the file descriptor
```

Comment:
I've added DEBUG logs to the code and I was able to depict the flow:

```
Uploading file
  golem-js:golem-network:storage [INFO] Preparing file upload { sourcePath: './tmp/test-refresh.txt' } +0ms
  golem-js:golem-network:storage [DEBUG] Received GFTP request for publishFile { component: 'GetMetadata', id: '7edfe3e8-a0a1-4adc-916a-d05d5aa92b16', payload: {} } +157ms
  golem-js:golem-network:storage [DEBUG] Received GFTP request for publishFile { component: 'GetChunk', id: '2bbcd2c4-a0d2-4b3d-9f24-3246be90d4bd', payload: { offset: 0, size: 40960 } } +51ms
  
Downloading file
  golem-js:golem-network:storage [INFO] Preparing file download { destination: './tmp/test-refresh.downloaded.txt' } +200ms
  golem-js:golem-network:storage [DEBUG] Received GFTP request for receiveFile { component: 'UploadChunk', id: '4a235443-90ed-466c-825c-d0394cd5cc5f', payload: { chunk: { content: [Uint8Array], offset: 0 } } } +158ms
  golem-js:golem-network:storage [DEBUG] Received GFTP request for receiveFile { component: 'UploadFinished', id: 'fe4e1f4f-1805-489c-b79c-ca5a8251bcc5', payload: { hash: '0f86628c2d9ba75beff1580d672a8d839c3e1dfaf6e50f4c0c9a5db071e70c7a' } } +48ms
```

It seems that we don't have a "finalization" message. I don't think one is needed since that's a "serving" mode and as a "server" we don't need to know if the client is done or not.

**To summarize:**

I think that leaving the "close" of the handles to the `close` method of the WebSocketStorageProvider class will be sufficient and we don't need that extra code in the finally block.

Reply:
Looking at [yagna source](https://github.com/golemfactory/yagna/blob/3d584097a0b2759d0184d9e89ea0756db1779bbb/core/model/src/gftp.rs) I see that there is no `DownloadFinished` model :/ 
Since we're only reading the files it's ok to have potential multiple open handles at the same time, I'll remove the closing logic from the message callback